cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```
@@ -71,6 +77,103 @@ public Schema getSchema() {
+   * @return value of the field as an object of {@link ZonedDateTime}, returns null if given field is not of
+   * logical type timestamp
+   */
+  public ZonedDateTime getTimestamp(String fieldName) {
+    if (!isValidLogicalType(fieldName, Schema.LogicalType.TIMESTAMP_MILLIS)
+      && !isValidLogicalType(fieldName, Schema.LogicalType.TIMESTAMP_MICROS)) {
+      return null;
+    }
+    return getZonedDateTime(fieldName, ZoneId.systemDefault());
+  }
+
+  /**
+   * Get {@link ZonedDateTime} from field of logical type timestamp.
+   * @param fieldName zoned date time field to get.
+   * @param zoneId zone id for the field
+   * @return value of the field as an object of {@link ZonedDateTime}
+   */
+  public ZonedDateTime getTimestamp(String fieldName, ZoneId zoneId) {
+    if (!isValidLogicalType(fieldName, Schema.LogicalType.TIMESTAMP_MILLIS)
+      && !isValidLogicalType(fieldName, Schema.LogicalType.TIMESTAMP_MICROS)) {
+      return null;
+    }
+
+    return getZonedDateTime(fieldName, zoneId);
+  }
+
+  /**
+   * Get zoned date and time represented by the field.
+   * @param fieldName Name of the field
+   * @param zoneId zone id for the field
+   * @return {@link ZonedDateTime} represented by field.
+   */
+  private ZonedDateTime getZonedDateTime(String fieldName, ZoneId zoneId) {
+    Long ts = (Long) fields.get(fieldName);
+    // get fractional portion of timestamp, we always store timestamp in micro seconds, so first get that portion.
+    // for example: timestamp is 1970-01-01T00:00:01.112456Z, here fraction = 112456, tsInSeconds = 1
+    int fraction = (int) (ts % (1000000L));
+    long tsInSeconds = (ts - fraction) / (1000000L);
+    // create an Instant with time in seconds and fraction which will be stored as nano seconds.
+    Instant instant = Instant.ofEpochSecond(tsInSeconds, fraction * 1000);
+    return ZonedDateTime.ofInstant(instant, zoneId);
+  }
+
+  private boolean isValidLogicalType(String fieldName, Schema.LogicalType logicalType) {
```

instead of returning whether it's valid, change this to a void validateLogicalType() call that throws exceptions with clear error messages.